### PR TITLE
[PGNCCL] Add default value for `nccl_nonblocking_timeout`

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138527
* #138384
* __->__ #138374
* #138488
* #137855

- Added default value for `nccl_nonblocking_timeout` (30 mins, previous: -1).
- Reuse C10D_CHECK_TIMEOUT in other CHECK macros

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/cuda/nccl.cpp [hunk 1]
```diff
@@ -155,40 +157,34 @@ bool nccl_use_nonblocking() {
...
-                             .count();
-      if (timeElapsed > nccl_nonblocking_timeout()) {
-        throw std::runtime_error("NCCL timeout.");
-      }
+    auto currentTimepoint = std::chrono::steady_clock::now();
+    auto timeElapsed = std::chrono::duration_cast<std::chrono::seconds>(
+                           currentTimepoint - startTimepoint)
+                           .count();
+    if (timeElapsed > nccl_nonblocking_timeout()) {
+      throw std::runtime_error("NCCL timeout.");
```
Comment (eqy) [inline torch/csrc/cuda/nccl.cpp:185]: For debugging purposes, could we indicate that this is triggered through the nonblocking path?

**Review by eqy — APPROVED** [review]

File: torch/csrc/cuda/nccl.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/cuda/nccl.cpp:185]: Good idea! I will add that!

File: torch/csrc/distributed/c10d/NCCLUtils.cpp [hunk 2]
```diff
@@ -179,26 +178,21 @@ bool nccl_use_nonblocking() {
...
-    timeout = std::stoi(config);
-    if (!nccl_use_nonblocking() && timeout > 0) {
-      TORCH_WARN(
-          "TORCH_NCCL_NONBLOCKING_TIMEOUT has no effect when TORCH_NCCL_USE_COMM_NONBLOCKING is false.");
-      timeout = -1;
+// Default value: 30 minutes
+int nccl_nonblocking_timeout() {
+  static int timeout = -2; // -2 means not initialized
+  if (timeout == -2) {
+    const char* val = getenv("TORCH_NCCL_NONBLOCKING_TIMEOUT");
```
Comment (cyyever) [inline torch/csrc/distributed/c10d/NCCLUtils.cpp:185]: Use our new c10::utils::get_env?

Comment (kwen2501) [reply torch/csrc/distributed/c10d/NCCLUtils.cpp:185]: Yes, I am using it [here](https://github.com/pytorch/pytorch/pull/138527/files#diff-c4e79e1dee3eea72b89d4471b91de8a7dae284e27bd5c32ed3e5296a561e4ee9) in a PR stacked on top :)
And the current piece of code is being deleted in that PR.
Thanks for pointing that out!

**Pull request closed without merging**